mat64/dense.go

```diff
@@ -497,3 +500,71 @@ func (m *Dense) Augment(a, b Matrix) {
 	w := m.View(0, ac, br, bc).(*Dense)
 	w.Copy(b)
 }
+
+// MarshalBinary encodes the receiver into a binary form and returns the result.
+//
+// Dense is little-endian encoded as follows:
+//  0 -  8  number of rows    (int64)
+//  8 - 16  number of columns (int64)
+// 16 - ..  matrix data elements (float64)
+//          [0,0] [0,1] ... [0,ncols-1]
+//          [1,0] [1,1] ... [1,ncols-1]
+//          ...
+//          [nrows-1,0] ... [nrows-1,ncols-1]
+func (m Dense) MarshalBinary() ([]byte, error) {
+	var err error
+	buf := bytes.NewBuffer(make([]byte, 0, m.mat.Rows*m.mat.Cols*8+16))
+	err = binary.Write(buf, defaultEndian, int64(m.mat.Rows))
+	if err != nil {
+		return nil, err
+	}
+
+	err = binary.Write(buf, defaultEndian, int64(m.mat.Cols))
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < m.mat.Rows; i++ {
+		for _, v := range m.rowView(i) {
+			err = binary.Write(buf, defaultEndian, v)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return buf.Bytes(), err
+}
+
+// UnmarshalBinary decodes the binary form into the receiver.
+//
+// See MarshalBinary for the on-disk layout.
+func (m *Dense) UnmarshalBinary(data []byte) error {
+	buf := bytes.NewReader(data)
+	var rows int64
+	err := binary.Read(buf, defaultEndian, &rows)
+	if err != nil {
+		return err
+	}
+
+	var cols int64
+	err = binary.Read(buf, defaultEndian, &cols)
+	if err != nil {
+		return err
+	}
+
+	m.mat.Rows = int(rows)
+	m.mat.Cols = int(cols)
+	m.mat.Stride = int(cols)
+	m.capRows = int(rows)
+	m.capCols = int(cols)
+	m.mat.Data = make([]float64, m.mat.Rows*m.mat.Cols)
```

Comment:
hum... I am not sure `zero` quite cuts it. it uses the `copy` builtin, which won't work in case `cap` is too small.
rather, one should check `cap(m.mat.Data)` for enough space, if there isn't, we'll need to re-allocate: better do that in one big chunk rather than discovering it too late in the game (ie: a big `make`)
otherwise, if `len(m.mat.Data)` is too small, `append` in one big chunk.
finally, `m.mat.Data = m.mat.Data[: m.mat.Rows*m.mat.Cols]` (and perhaps also set the cap?)

besides, all the data will be overwritten by the for-loop and `binary.Read` so the for-loop of `zero` is a bit superfluous if you ask me.

so:

```go
sz := m.mat.Rows * m.mat.Cols
if cap(m.mat.Data) < sz {
   m.mat.Data = make([]float64, sz)
}
if len(m.mat.Data) < sz {
  m.mat.Data = append(m.mat.Data, make([]float64, sz-len(m.mat.Data))...)
}
if len(m.mat.Data) > sz {
  m.mat.Data = m.mat.Data[:sz]
  // or, even:
  // m.mat.Data = m.mat.Data[:sz:sz]
}
```

to be completely honest, I am not sure all of these (premature?) optimizations are worth it.

in 90% of cases, `UnmarshalBinary` will be used on just-created `Dense` matrices... won't it?